_posts/2016-08-02-testing-spark-streaming-unit-testing.md

```diff
@@ -0,0 +1,126 @@
+- Stateful operations: local or remote state.
+- Latency: real time or near real time.
+- Reliability, high availability and durability.
+
+## What is Spark Streaming
+
+Spark has been a revolution in the [Big Data](https://www.wikiwand.com/en/Big_data) space. It has replaced [Hadoop's MapReduce](https://hadoop.apache.org/docs/current/hadoop-mapreduce-client/hadoop-mapreduce-client-core/MapReduceTutorial.html) as the preferred batch processing framework. The main reasons are:
+
+- Speed: [Run programs up to 100x faster than Hadoop MapReduce in memory, or 10x faster on disk.](http://spark.apache.org/)
+- Usability: MapReduce DSL is far from easy to write and read. Spark Scala DSL works as an extension of Scala Collections operations so the learning curve is really small.
+- Community: there are a lot of excitement around Spark and there are plenty of related tools like MLib, Spark SQL or Spark Streaming.
+
+Spark Streaming is built on top of Spark. That means that you can use Spark infrastructure and concepts such as [YARN](http://hadoop.apache.org/docs/current/hadoop-yarn/hadoop-yarn-site/YARN.html), [HDFS](https://hadoop.apache.org/docs/r1.2.1/hdfs_design.html) or [RDDs](http://www.tutorialspoint.com/apache_spark/apache_spark_rdd.htm). On top of that we'll have abstractions to help us building streaming features like aggregations or sliding windows.
+
+## What is Unit testing
+
+[This](http://robdmoore.id.au/blog/2015/01/26/testing-i-dont-even/) is a fantastic series about different views on unit testing. To keep the scope of this post focused, we'll work with the following characteristics.
+
+- Network Isolation: production code under test will involve code that lives in a single process. No network calls are allowed.
+- Framework Isolation: we want to test our code, as much as possible, and not the interactions with underlying frameworks.
+
+## Spark Testing Base to the rescue
+
+Controlling the lifecycle of Spark can be cumbersome and tedious. Fortunately, [Spark Testing Base](https://github.com/holdenk/spark-testing-base) project offers us Scala Traits that handle those low level details for us. Streaming has an extra bit of complexity as we need to produce data for ingestion in a timely way. At the same time, Spark internal clock needs to tick in a controlled way if we want to test timed operations as sliding windows.
+
+Let's see how to test the archetypical example of WordCount:
+
+```scala
+def count(lines: DStream[String]): DStream[(String, Int)] =
+ lines.flatMap(_.split(" "))
+   .map(word => (word, 1))
+   .reduceByKey(_ + _)
+```
+
+As you can see, this is pure function, with no side effects or access to external state. We can reason about it by having a look to the signature of the function. [DStream](https://spark.apache.org/docs/0.7.2/api/streaming/spark/streaming/DStream.html) is the basic abstraction in Spark Streaming and Spark Testing Base will help us to deal with it.
```

Comment:
a look to the signature of the function ==> a look `at` the signature of the function